parsers/CA_NS.py

```diff
@@ -1,212 +1,179 @@
-    imports = []
-    for mix in mix_data:
-        percent_mix = {
-            "coal": mix["Solid Fuel"] / 100.0,
-            "gas": (mix["HFO/Natural Gas"] + mix["LM 6000's"]) / 100.0,
-            "oil": mix["CT's"] / 100.0,
-            "biomass": mix["Biomass"] / 100.0,
-            "hydro": mix["Hydro"] / 100.0,
-            "wind": mix["Wind"] / 100.0,
-            "imports": mix["Imports"] / 100.0,
-        }
-
-        # datetime is in format '/Date(1493924400000)/'
-        # get the timestamp 1493924400 (cutting out last three zeros as well)
-        data_timestamp = int(mix["datetime"][6:-5])
-        data_date = datetime.fromtimestamp(data_timestamp, tz=timezone.utc)
-
-        # validate
-        valid = True
-        for gen_type, value in percent_mix.items():
-            percent_bounds = valid_percent[gen_type]
-            if not (percent_bounds[0] <= value <= percent_bounds[1]):
-                # skip this datapoint in the loop
-                valid = False
-                logger.warning(
-                    f"discarding datapoint at {data_date} due to {gen_type} percentage "
-                    f"out of bounds: {value}",
-                    extra={"key": zone_key},
-                )
-        if not valid:
-            # continue the outer loop, not the inner
-            continue
+    if zone_key != ZONE_KEY:
+        raise ParserException(PARSER, f"Cannot parse zone '{zone_key}'", zone_key)
 
-        # in mix_data, the values are expressed as percentages,
-        # and have to be multiplied by load to find the actual MW value.
-        corresponding_load = [
-            load_period
-            for load_period in load_data
-            if load_period["datetime"] == mix["datetime"]
-        ]
-        if corresponding_load:
-            load = corresponding_load[0]["Base Load"]
+    # Request data from the source. Skip the first element of each JSON array
+    # because the reported base load is always 0 MW.
+    session = session or Session()
+    loads = {  # A lookup table mapping timestamps to base loads (in MW)
+        _parse_timestamp(load["datetime"]): load["Base Load"]
+        for load in session.get(LOAD_URL).json()[1:]
+    }
+    mixes = session.get(MIX_URL).json()[1:]  # Electricity mix breakdowns in %
+
+    production_breakdowns = ProductionBreakdownList(logger)
+    for mix in mixes:
+        timestamp = _parse_timestamp(mix["datetime"])
+        if timestamp in loads:
+            load = loads[timestamp]
         else:
-            # if not found, assume 1244 MW, based on average yearly electricity available for use
-            # in 2014 and 2015 (Statistics Canada table Table 127-0008 for Nova Scotia)
+            # If a base load corresponding with this timestamp is not found,
+            # assume 1244 MW based on the average yearly electricity available
+            # for use in 2014 and 2015 (Statistics Canada table 127-0008 for
+            # Nova Scotia).
             load = 1244
```

Comment:
We should remove this here because we are generating synthetic data and we don't keep track of it. It is better to have a None that will later be estimated

Reply:
for the record, this was first added in https://github.com/electricitymaps/electricitymaps-contrib/pull/503 back in 2017

Reply:
yes, i know. We are trying to improve the overall quality of our data and this means that we want to minimize creating data points at the parser level and we want to ingest as little synthetic data as possible. If these points are missing, they will be estimated by our pipelines which will detect the gap in the time series

Reply:
When the base load is `None`, every mode in the `ProductionMix` object should be set to `None` because each one is derived from the base load. In that case, can we just skip the offending data point / event entirely?

Reply:
> yes, i know. We are trying to improve the overall quality of our data and this means that we want to minimize creating data points at the parser level and we want to ingest as little synthetic data as possible. If these points are missing, they will be estimated by our pipelines which will detect the gap in the time series

FYI, if you want to correct existing data, NS power has [historical data on load from 2012 on](https://www.nspower.ca/oasis/monthly-reports/hourly-total-net-nova-scotia-load). So it should be possible to correct all existing datapoints if you want to.

Reply:
that's great to know, this could be useful to fill in the missing data!

@kruschk when all modes are None the parser validation layer will filter out the points that have completely empty (all `None`) so there is nothing to be done in the parser actually